go-controller/pkg/cni/cnismartnic.go

```diff
@@ -0,0 +1,114 @@
+package cni
+
+// contains code for cnishim - one that gets called as the cni Plugin
+// This does not do the real cni work. This is just the client to the cniserver
+// that does the real work.
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/Mellanox/sriovnet"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// Plugin is the structure to hold the endpoint information and the corresponding
+// functions to use it
+type SmartNicPlugin struct {
+}
+
+// NewCNISmartNicPlugin creates the internal Plugin object
+func NewCNISmartNicPlugin() *SmartNicPlugin {
+	return &SmartNicPlugin{}
+}
+
+// CmdAdd is the callback for 'add' cni calls from skel
+func (bp *SmartNicPlugin) CmdAdd(args *skel.CmdArgs) error {
+	// read the config stdin args to obtain cniVersion
+	conf, err := config.ReadCNIConfig(args.StdinData)
+	if err != nil {
+		return fmt.Errorf("invalid stdin args")
+	}
+
+	req := newCNIRequest(args)
+	podReq, err := cniRequestToPodRequest(req)
+	if err != nil {
+		return fmt.Errorf("cniRequestToPodRequest failed %v", err)
+	}
+
+	// 1- Verify there is a device id
+	if podReq.CNIConf.DeviceID == "" {
+		return fmt.Errorf("DeviceID must be set")
+	}
+	pciAddress := podReq.CNIConf.DeviceID
+
+	// 2- Get the PF index and VF index
+	pfPciAddress, err := GetPfPciFromVfPci(pciAddress)
+	if err != nil {
+		return err
+	}
+	pfindex := pfPciAddress[len(pfPciAddress)-1]
+	vfindex, err := sriovnet.GetVfIndexByPciAddress(pciAddress)
+	if err != nil {
+		return err
+	}
+
+	// 3- Create kubernetes client
+	kConf, err := clientcmd.BuildConfigFromFlags("", conf.Kubeconfig)
+	if err != nil {
+		return fmt.Errorf("unable to set up client config error %v", err)
+	}
+	clientset, err := kubernetes.NewForConfig(kConf)
+	if err != nil {
+		return fmt.Errorf("unable to create a kubernetes client error %v", err)
+	}
+	kube := &kube.Kube{KClient: clientset}
+
+	// 4. Set smart-nic pod annotation of PF index and VF index
+	err = kube.SetAnnotationsOnPod(podReq.PodNamespace, podReq.PodName, map[string]string{
+		"ovn.smartnic.pf": string(pfindex), "ovn.smartnic.vf": strconv.Itoa(vfindex), "sandbox": podReq.SandboxID})
```

Comment:
please align according to to the annotations in the design document

https://docs.google.com/document/d/11IoMKiohK7hIyIE36FJmwJv46DEBx52a4fqvrpCBBcg/edit?usp=sharing